packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
+    _maybeAddCompletedFrame(f);
+
+    return eventAdded;
   }
 
-  TimelineEventType _getTimelineEventType(TraceEvent event) {
-    // The event will only be of type 'cpu' or 'gpu' because we do not process
-    // events of any other type.
-    if (event.isCpuEvent) {
-      return TimelineEventType.cpu;
-    } else {
-      return TimelineEventType.gpu;
+  bool _eventOccursWithinFrameBoundaries(TimelineEvent e, TimelineFrame f) {
+    // Epsilon in microseconds.
+    const int epsilon = 50;
+
+    // Allow the event to extend the frame boundaries by `epsilon` microseconds.
+    final bool fitsStartBoundary = f.startTime - e.startTime - epsilon < 0;
+    final bool fitsEndBoundary = f.endTime - e.endTime + epsilon > 0;
+    return fitsStartBoundary && fitsEndBoundary;
+  }
+
+  List<TimelineFrame> _getAndSortWellFormedFrames() {
+    final List<TimelineFrame> frames =
+        List<TimelineFrame>.from(_pendingFrames.values)
+            .where((TimelineFrame f) => f.wellFormed)
+            .toList();
+
+    // Sort frames by their startTime. Sorting these frames ensures we will
+    // handle the oldest frame first when iterating through the list.
+    frames.sort((a, b) => a.startTime.compareTo(b.startTime));
+    return frames;
+  }
+
+  void _maybeAddCompletedFrame(TimelineFrame frame) {
+    if (frame.readyForTimeline && !frame.addedToTimeline.isCompleted) {
+      _frameCompleteController.add(frame);
+      _pendingFrames.remove(frame);
+      frame.addedToTimeline.complete();
     }
   }
 }
 
+/// Data describing a single frame.
+///
+/// Each TimelineFrame should have 4 distinct pieces of data:
+///   - [_pipelineProduceFlow] : flow of events showing the CPU work for the
+///     frame.
+///   - [_pipelineConsumeFlow] : flow of events showing the GPU work for the
+///     frame.
 class TimelineFrame {
-  TimelineFrame(this.id, this.startTime);
+  TimelineFrame(this.id);
 
   final String id;
 
-  final List<TimelineEvent> cpuEvents = [];
-  final List<TimelineEvent> gpuEvents = [];
+  /// Marks whether this frame has been added to the timeline.
+  Completer<Null> addedToTimeline = Completer();
+
+  /// Flow of events showing the CPU work for the frame.
+  TimelineEvent get pipelineProduceFlow => _pipelineProduceFlow;
+
+  /// Flow of events showing the GPU work for the frame.
+  TimelineEvent get pipelineConsumeFlow => _pipelineConsumeFlow;
+
+  TimelineEvent _pipelineProduceFlow;
+  TimelineEvent _pipelineConsumeFlow;
+
+  set pipelineProduceFlow(TimelineEvent e) {
+    assert(_pipelineProduceFlow == null, 'pipelineProduceFlow already set');
+    _pipelineProduceFlow = e;
+  }
+
+  set pipelineConsumeFlow(TimelineEvent e) {
+    assert(_pipelineConsumeFlow == null, 'pipelineConsumeFlow already set');
+    _pipelineConsumeFlow = e;
+  }
+
+  /// Whether the frame is ready for the timeline.
+  ///
+  /// A frame is ready once it has both required event flows as well as
+  /// [startTime] and [endTime].
+  bool get readyForTimeline {
```

Comment:
nit: make bools like this have is in the name
e.g. 
isReady
or
isReadyForTimeline.

Reply:
done.